eval/src/vespa/eval/eval/fast_value.hpp

```diff
@@ -360,4 +367,45 @@ FastValueIndex::sparse_only_merge(const ValueType &res_type, const Fun &fun,
 
 //-----------------------------------------------------------------------------
 
+template <typename LCT, typename RCT, typename OCT, typename Fun>
+const Value &
+FastValueIndex::sparse_no_overlap_join(const ValueType &res_type, const Fun &fun,
+                                       const FastValueIndex &lhs, const FastValueIndex &rhs,
+                                       const std::vector<JoinAddrSource> &addr_sources,
+                                       ConstArrayRef<LCT> lhs_cells, ConstArrayRef<RCT> rhs_cells, Stash &stash)
+{
+    using HashedLabelRef = std::reference_wrapper<const FastSparseMap::HashedLabel>;
+    auto &result = stash.create<FastValue<OCT>>(res_type, res_type.count_mapped_dimensions(), 1, lhs.map.size()*rhs.map.size());
+    std::vector<HashedLabelRef> output_addr;
+    for (size_t lhs_subspace = 0; lhs_subspace < lhs.map.size(); ++lhs_subspace) {
+        auto l_addr = lhs.map.make_addr(lhs_subspace);
+        for (size_t rhs_subspace = 0; rhs_subspace < rhs.map.size(); ++rhs_subspace) {
+            auto r_addr = rhs.map.make_addr(rhs_subspace);
+            output_addr.clear();
+            size_t l_idx = 0;
+            size_t r_idx = 0;
+            for (JoinAddrSource source : addr_sources) {
+                switch (source) {
+                case JoinAddrSource::LHS:
+                    output_addr.push_back(l_addr[l_idx++]);
```

Comment:
we should be able to avoid having to handle the lhs address elements again for each rhs element...